pennylane/interfaces/jax_jit_tuple.py

```diff
@@ -382,40 +366,78 @@ def wrapper(p):
+        for i, (new_trainable_params, tape) in enumerate(zip(list_new_trainable_parameters, tapes)):
+            if tape.trainable_params != new_trainable_params:
+                tape.trainable_params = new_trainable_params
+                switch_trainable[i] = True
+
+        # Forward execution with the right trainable parameters
         res, jacs = execute_wrapper(primals[0])
+
+        jacs_ = [jacs] if len(tapes) == 1 else jacs
         multi_measurements = [len(tape.measurements) > 1 for tape in tapes]
 
-        if len(tapes) == 1:
-            jacs = [jacs]
+        updated_jacs = []
 
-        jvps = _jit_compute_jvps(jacs, tangents[0], multi_measurements, trainable_params)
-        return res, jvps
+        # Add zeros in the jacobians if the trainable params were switched
+        for i, (switch, multi_measurement) in enumerate(zip(switch_trainable, multi_measurements)):
+            if switch:
+                intermediate_jacs = []
 
-    res = execute_wrapper(params)
+                # Adapt the shape of the empty jacobian given the measurement shape
+                shape_dtype = original_shape[0] if len(tapes) == 1 else original_shape[i][0]
 
-    tracing = []
-    for i, tape in enumerate(tapes):
-        if len(tape.measurements) == 1:
-            tracing.append(isinstance(res[i], jax.interpreters.ad.JVPTracer))
-        else:
-            tracing.append(any(isinstance(r, jax.interpreters.ad.JVPTracer) for r in res[i]))
+                # Multi measurement
+                if multi_measurement:
+                    jac_empty = tuple(
+                        jax.numpy.zeros(shape=tensor.shape, dtype=tensor.dtype)
+                        for tensor in shape_dtype
+                    )
+                # Single measurement
+                else:
+                    jac_empty = jnp.zeros(shape_dtype.shape, shape_dtype.dtype)
+
+                for j, original_trainable_params in enumerate(
+                    list_original_trainable_parameters[i]
+                ):
+                    if original_trainable_params in list_new_trainable_parameters[i]:
+                        intermediate_jacs.append(jacs_[i])
+                    else:
+                        intermediate_jacs.append(jac_empty)
+                updated_jacs.append(tuple(intermediate_jacs))
 
-    tracing = any(tracing)
+        if updated_jacs:
+            updated_jacs = updated_jacs[0] if len(tapes) == 1 else tuple(updated_jacs)
 
-    # When there are no tracers (not differentiating), we have the result of
-    # the forward pass and the jacobian, but only need the result of the
-    # forward pass
-    if len(res) == 2 and not tracing:
-        res = res[0]
+        # Get the jvps
+        tangents = _filter_zeros_tangents(tangents[0])
+        jvps = _compute_jvps(jacs_, tangents, multi_measurements)
+
+        jacs = jacs if not updated_jacs else updated_jacs
+        return (res, jacs), (jvps, jacs)
```

Comment:
It looks like the `jacs` returned from this function will only contain jacobians for the tapes with `switch=True`. What happens if there are other tapes in the batch with `switch=False`?

Reply:
Sorry, I'm still not understanding. If for example we have two tapes and `switch_trainable = [True, False]`, then `updated_jacs` will only contain the jacobians for the first tape. So in this case we only return jacobians for the first tape. Don't we want to return the jacobians for both tapes?

Reply:
Yes I think you are right, let me update that :+1: